cc/dual_net/factory.cc

```diff
@@ -50,56 +61,76 @@ DEFINE_string(engine, MG_DEFAULT_ENGINE,
 #ifdef MG_ENABLE_LITE_DUAL_NET
               " \"lite\""
 #endif
+#ifdef MG_ENABLE_TPU_DUAL_NET
+              " \"tpu\""
+#endif
 #ifdef MG_ENABLE_TRT_DUAL_NET
               " \"trt\""
 #endif
 );
 
+// TPU flags.
+DEFINE_string(
+    tpu_name, "",
+    "Cloud TPU name to run inference on, e.g. \"grpc://10.240.2.10:8470\"");
+
 // Batching flags.
 // TODO(tommadams): Calculate batch_size from other arguments (virtual_lossess,
 // parallel_games, etc) instead of having an explicit flag.
 DEFINE_int32(batch_size, 256, "Inference batch size.");
 
 namespace minigo {
-namespace {
 
-std::unique_ptr<DualNet> NewDualNet(const std::string& graph_path) {
+DualNetFactory::~DualNetFactory() = default;
+
+std::unique_ptr<DualNetFactory> NewDualNetFactory(
+    const std::string& model_path) {
+  std::unique_ptr<DualNet> dual_net;
+  int batch_size = FLAGS_batch_size;
+
   if (FLAGS_engine == "tf") {
 #ifdef MG_ENABLE_TF_DUAL_NET
-    return NewTfDualNet(graph_path);
+    dual_net = NewTfDualNet(model_path);
 #else
     MG_FATAL() << "Binary wasn't compiled with tf inference support";
 #endif  // MG_ENABLE_TF_DUAL_NET
   }
 
   if (FLAGS_engine == "lite") {
 #ifdef MG_ENABLE_LITE_DUAL_NET
-    return NewLiteDualNet(graph_path);
+    dual_net = NewLiteDualNet(model_path);
 #else
     MG_FATAL() << "Binary wasn't compiled with lite inference support";
 #endif  // MG_ENABLE_LITE_DUAL_NET
   }
 
+  if (FLAGS_engine == "tpu") {
+#ifdef MG_ENABLE_TPU_DUAL_NET
+    int inferences = FLAGS_virtual_losses * FLAGS_parallel_games;
+    int buffering = FLAGS_parallel_games == 1 ? 1 : 2;
```

Comment:
i don't think this is needed -- you can double buffer on a single game if you've got vlosses = 2* inference batch size.

Reply:
done